src/curl_util.cpp

```diff
@@ -256,8 +256,13 @@ std::string prepare_url(const char* url)
     std::string uri;
     std::string hostname;
     std::string path;
+    std::string token;
     std::string url_str = std::string(url);
-    std::string token = std::string("/") + bucket;
+    if(is_multi_tenancy) {
+      token = std::string("/") + urlEncode(bucket);
```

Comment:
I don't understand this change -- can we always urlEncode the bucket name?  What are the valid characters?  How do other S3 clients handle this?

Reply:
Url (containing bucket) is passed to the function as urlEncoded, so the token used to separate hostname and path need to be urlencoded as well. I guess bucket could be urlEncoded always, but I wanted to keep changes minimum. 

For example [AWSJavaSDK doc](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/AmazonS3URI.html
) says: 
> URL encoding is recommended if you work with bucket names or object keys with special characters. This can be disabled if you pre-encode URI strings before passing them to this class.

Reply:
The argument `url` of this prepare_url function is URL encoded by MakeUrlResource function, so if we don't encode `token` as well, we may not be able to match.
In other words, it seems that the bucket name required by CEPH contains `:`, which is encoded.

It should be noted that except for path style, the domain name will contain `:`, so I think that it will not be processed correctly as FQDN. (Even if encoded, it will contain `%`)

Wouldn't it be necessary to enforce the path style to get this CEPH support?

Reply:
[Ceph documentation](https://docs.ceph.com/en/latest/radosgw/multitenancy/) says:
>Note that it’s not possible to supply an explicit tenant using a hostname. Hostnames cannot contain colons, or any other separators that are not already valid in bucket names. Using a period creates an ambiguous syntax. Therefore, the bucket-in-URL-path format has to be used.

So, forcing path style when CEPH multi tenancy naming convention is used makes sense.

Committed corresponding changes to the pull request.

Reply:
Thank you, I've confirmed your code fix that checks for a combination of is_multi_tenancy and pathrequeststyle.

@gaul
I think this fix is fine, please give us your opinion.

Reply:
What are other clients doing with this multi-tenant option?  Do they have an explicit option, always URL-encode the bucket name, or require the user to manually URL-encode the `:`?  I am concerned about adding even more flags unless they are necessary.

Reply:
> What are other clients doing with this multi-tenant option? [...] I am concerned about adding even more flags unless they are necessary.
Well, this is a CEPH functionality so I assume the need to support this is general.

> Do they have an explicit option, always URL-encode the bucket name, or require the user to manually URL-encode the :?
Not sure if I understood this question, but user do no tneed to url-encode the :

Reply:
When I look at the [Ceph example](https://docs.ceph.com/en/latest/radosgw/multitenancy/#s3) it doesn't have a special tenancy id option -- it includes the tenant in the bucket.  Thus I believe that s3fs users should also specify the tenant in the bucket and if there is an encoding issue it should be part of the bucket and not some weird option that only s3fs has.